web/src/app/validation/ceg/contradictory-condition-validator.ts

```diff
@@ -0,0 +1,285 @@
+            let toNode = this.topoNodes[to];
+
+            if (this.isAndNode[to]) {
+                /**
+                 * We have an AND Node:
+                 * All parent causes are also causes of this node.
+                 * --> Create the union of all parent causes.
+                 */
+                toNode.edgeUnion(tNode, edge);
+            } else {
+                /**
+                 * We have an OR Node:
+                 * Only causes, that are necessary causes of ALL parents are also causes of this node.
+                 * --> Create the intersection of all parent causes.
+                 */
+                toNode.edgeIntersection(tNode, edge);
+            }
+            // We have dealt with this node.
+            // Reduce the parent count (Topological Sort)
+            toNode.parentCount--;
+            if (toNode.parentCount == 0) {
+                // We have explored all parents of toNode. It can now be explored.
+                // Test for contraditcions and add mark it for the openList.
+                for (const key in toNode.posEdgeSets) {
+                    if (toNode.posEdgeSets.hasOwnProperty(key) && toNode.negEdgeSets.hasOwnProperty(key)) {
+                        // We have a contradiction
+                        // Cause -PosPath-> Effect
+                        // Cause -NegPath-> !Effect
+                        let cause = this.topoNodes[key].node;
+                        let effect = toNode.node;
+                        let posPath = toNode.posEdgeSets[key];
+                        let negPath = toNode.negEdgeSets[key];
+                        let contradiction = new Contradiction(cause, effect, posPath, negPath);
+                        contradictions.push(contradiction);
+                    }
+                }
+                newUnexploredNodes.push(toNode);
+            }
+        }
+        return [newUnexploredNodes, contradictions];
+    }
+}
+
+class TopologicalNode {
+    /**
+     * Helper Datastructure:
+     * A CEG Node with all its outgoing edges.
+     * A counter for the number of unexplored parents (for the topological sorting)
+     * and the nodes' positive and and negative cause sets.
+     */
+    public node: CEGNode;
+    public edges: CEGConnection[];
+    public parentCount: number;
+    public posEdgeSets: {[parentURL: string]: CEGConnection[]};
+    public negEdgeSets: {[parentURL: string]: CEGConnection[]};
+    private _setWasFilled: boolean;
+
+    constructor(node: CEGNode) {
+        this.node = node;
+        this.edges = [];
+        this.parentCount = 0;
+        this.posEdgeSets = {};
+        this.negEdgeSets = {};
+        this._setWasFilled = false;
+    }
+
+    private static cloneEdgeSet(edgeSet: {[parentURL: string]: CEGConnection[]}): {[parentURL: string]: CEGConnection[]} {
+        /**
+         * Helper function. Creates copy of an edge set. Also copys the arrays within the set.
+         */
+        let out = {};
```

Comment:
Also here, you can use `const` instead of `let` - please review your code regarding the use of `let` and `const`.